### ISSUE TITLE
Implement schema update logic

### ISSUE THREAD

**Issue by tssgery** [RaspberryPints/RaspberryPints#213]
I am working on some logic to allow the upgrading of the db schema and am seeking comments

Added files (so far)

```
/includes/check_database.php
/sql/updates/*.sql
```

The ways this work is:
- The /includes/check_database.php script is included from both /index.php as well as /admin/index.php, which cause it to check the database on every refresh. Since this is a low volume webapp the additional include file should not cause a negative impact on performance.
- Within check_database.php, we get the current schema version from the database and compare it to the version that we need to have (hard coded in check_database.php right now). If the versions do not match, all of the sql files in the sql/update directory are examined to see if they need to be run. For each sql file found, if the version number is greater than the version we have and less than (or equal to) the version we need... we run the sql file and update the current database schema version.
- Version numbers for schema are in the form of YYYYMMDD-HHMMSS.
- SQL update files are named in a similar convention, YYYYMMDD-HHMMSS-something.sql, which allows us to sort the sql names and process them in order.

The pseudocode for check_databse.php looks like

```
get current version from database
get needed versions
if (needed != current) {
    foreach file (in a sorted list) that is newer than what we have and less than/equal to what we need {
         process the sql
         update current version in database  
     }
}
```

**Comment by tssgery** [comment]
I have started implementing this. If it helps to look at the work in progress, it is in my updatedb branch of my fork: https://github.com/tssgery/RaspberryPints/tree/updatedb

**Comment by kerber** [comment]
I'll have to take some time to dig into this later, but this looks like a great avenue for schema updates! I'll be excited to hear everyone else's comments. 
A couple questions/thoughts:
- Any plans to have it backup the DB before running changes?
- I could see this working well as part of the upgrade process. New files are pulled down over the current ones and the page is reloaded. On reload the script sees the DB needs to be upgraded and kicks it off.

**Comment by tssgery** [comment]
You are correct. It should really backup the database before starting any upgrades. I assume that it's ok if, on failure, restoring the database would be a manual operation?

**Comment by JayUnt** [comment]
Sounds great.

How are the versions associated to the date named files? In other words, hour do you know that for version 1.2.8 you should run that file?

**Comment by JayUnt** [comment]
Nm I see you said the sql versions are named with the dates. Then how are we associating the code version with the schema version?

**Comment by tssgery** [comment]
Correct. With this scheme, the database schema is versioned by date while the product can be versioned by traditional number. The mapping between code version and schema version will need to be done manually. That's not as bad as it sounds though. As an example, if you plan to release version 1.1 on 3/15/2014... you could go into the check_database.php file today and set the version needed to 20140315_235959 and all schema updates up to then will automatically be included.

I considered versioning the schema the same as the code but wanted to allow multiple schema updates to be applied for any version upgrade. With the schema being version by date, developers can/should name the sql files appropriately as the code is being developed without worrying about what version of the code they will be needed by.

**Comment by tssgery** [comment]
additionally, I should mentioned that if you want all schema updates on the branch to be applied and not worry about the "version needed"... set the needed version to a date way in the future, 20991231_235959 or similar. Then every update would be applied

**Comment by JayUnt** [comment]
So then when a new version is decided on, we should make a blank sql file
for all updates for that version, right?   That way if I was to want to add
a change, it would already be there and I wouldn't have to figure out the
release date.  That should be added to the process of doing the release,
make the next releases sql file.

What if we don't know the next release's date? Would/could that get
confusing?

**Comment by tssgery** [comment]
That's not quite the process I had in mind. Here is what I was thinking (dates and versions are obviously made up):
- February 20: Version 1.0 is released
- March 1: Bob starts working on cool feature called 'fubar'. This feature needs a schema change so Bob creates a sql file named 20140301_080000-fubar.sql
- March 2: Steve realizes he needs a schema change for his 'uber' feature and creates 2014_0302_150000-uber.sql
- March 13: George fixes a defect which needs a sql change, creating 20140313_000000-dangbugs.sql
- March 20: Version 1.1 ships, with all three sql files and a required schema version of 20140313_000000 (or 20140320_235959). When the check_database.php script sees the new schema version needed, it processes all three sql files in order from oldest to newest.

This way, the schema changes get developed with features, normally in a branch, and merged to the release branch along with the feature. Files are named as they are created without worry of if/when it will ship.

**Comment by thadius856** [comment]
Good talks going on here.

Unfortunately, they're a bit over my head and I don't have the time this week to hit the pre-req reading. :\

I trust whatever you guys decide will be good stuff.

**Comment by tssgery** [comment]
BTW, we could easily change this to apply any missing schema updates instead of specify a needed version. The code then would simply loop though the sql scripts and run all of them it finds that it hasn't already processed.

EDIT: Just tried setting the needed version to "LATEST" and it did not work as expected. We can certainly remove the logic to stop updating at a specific version though if that's deemed better.

**Comment by kerber** [comment]
Merging before pushing to develop is going to be very tricky. 

Say Steve has been working on his code for two weeks and has his schema working and tested. He pulls develop and merges without conflicts. Awesome. He pushed. Suddenly, everyone's update is failing. Turns out he had been operating on a column that no longer exists because of Bob's feature merge.

While there won't be any conflicting code to be flagged by git, your previously working update might not work when merged. We'll have to add a step to our merge process to apply all schema updates to a fresh develop database directly before a merge and push.

**Comment by tssgery** [comment]
You are correct that people will have to add a step to test that applying all the database updates work properly, but I don't see a way around that. Any thoughts on how to avoid it?

**Comment by JayUnt** [comment]
This makes sense to me.  Nice work.

One way around that scenario is to have a just one branch per release. So when you commit your new code, you are committing to what is the current code base.  This obviously, has its own issues.

Another way, which is what we do at my work, is before the branch can be merged into the master, the developer of the branch, merges the master back into his branch.  He then resolves any conflicts and then can merge its all back into master.  That way, if there are a bunch of conflicts and things need to be discussed, its just broken on the branch and it doesn't hold the other developers up.  I personally like this method, because it puts it on the developer, who wrote the code, to resolve any conflicts.  They should have to do a rebase that comes in perfectly clean before merging back into the main branch or master or where it needs to get to.

**Comment by mcangeli** [comment]
I think @JayUnt has it with the second option there... thats what I was doing (some times unsuccessfully) with the fork I had created with the code for untappd.

**Comment by tssgery** [comment]
I agree with that approach as well and that's how we also handle it at work. Before merging to 'master', we merge 'master' into our feature branch and test. Once the tests are successful, we merge to 'master'. The good thing about this too, is that the result merge to 'master' almost always becomes trivial as the conflicts were resolved on the rebase.

**Comment by tssgery** [comment]
I'll leave this open for comments though EOD Friday, Feb 28. After that, I'll get back to implementing it so we can get a pull request going asap.

**Comment by thadius856** [comment]
> Another way, which is what we do at my work, is before the branch can be merged into the master, the developer of the branch, merges the master back into his branch. He then resolves any conflicts and then can merge its all back into master. That way, if there are a bunch of conflicts and things need to be discussed, its just broken on the branch and it doesn't hold the other developers up. I personally like this method, because it puts it on the developer, who wrote the code, to resolve any conflicts. They should have to do a rebase that comes in perfectly clean before merging back into the main branch or master or where it needs to get to.

I like this, with one caveat.

After merging master into develop, we should really be pushing develop to a new release candidate branch. This allows our beta testers to play with a version that is not constantly changing  on them. It also frees up the develop branch to continue along while the release candidate branch is vetted.

On release day, you merge the rc into master (to create the release), and also merge rc back into develop (to incorporate any fixes inspired by the beta testers).

Like so:

![screen-shot-2009-12-24-at-11 32 03](https://f.cloud.github.com/assets/6320180/2291027/04d098fe-a031-11e3-9a1f-39fe2a933ac8.png)

**Comment by JayUnt** [comment]
Looks good to me.  the only thing we should add is a arrow from the develop branch into the feature branch right before the feature branch gets merged into develop.  This is where the feature branch developer will resolve the conflicts, like changes made by a hotfix, or a different feature branch.  And the same with any other re-mergings, like hotfix into master.

We should also put this in a wiki page. So if any new developer comes, we can show them our process in the wiki.

**Comment by thadius856** [comment]
Yes, we should be merging the changes back into each feature.

I don't know if it exists in svn, but git calls this a rebase. It's pretty powerful, as it essentially re-writes the history of a branch based on another branch.

A couple of the tuts out there are pretty wordy. This is the one that made sense to me:

https://www.atlassian.com/git/tutorial/rewriting-git-history#!rebase

**Comment by JayUnt** [comment]
I just thought of a hiccup for this versioning using the date.  What if we do a release on 1/31/204, and then i make a branch off of that release for a new large feature.  I then start with a database update on 2/1/2014, so i create the file with that date. 

Then on 2/5/2014 we find a bug in the release so we do a hotfix, and make the file name for that schema using 2/5.  We push it into the release and its all good.

But now there is db scripts that is not released but has the date 2/1, and the last release (the hotfix) was on 2/5.  Would mine get skipped over?

**Comment by tssgery** [comment]
actually, yes it would get skipped over. Good catch.

We could change the logic to make sure that every sql file has been processed, regardless of the version/date but that leads to other issues (for example, your 2/1/2014 version is based upon updating 1/31/2014, while upon release it would really need to be based upon the 2/5/2014 schema... if we blindly process it, it could break things).

I'd rather not make people have to go and rename all their sql updates whenever we do a release. That's error prone and very ugly.

I have to noodle on this a bit.

**Comment by kerber** [comment]
This goes right along with the issue I'd mentioned earlier - If someone has a branch that's been off of develop for a while. There are almost certainly going to be other changes to the develop DB since they first branched. The last step of the pre-merge process should be to update their schema filename to the merge date. They also need to rebuild a database from the develop branch and run their change set against that. That is the only way to make sure that the person merging has accounted for all changes since they branched.

**Comment by kerber** [comment]
It's going to be a pain keeping branch DB changesets up to date and working, but I'm not sure how else without something overly complicated like liquibase.

**Comment by kerber** [comment]
We will have to also pay a lot of attention when merging hotfixes. Hotfix updates are going to be based off of the release. There's a chance that the develop DB could look very different and the hotfix update could fail or completely FUBAR it.

**Comment by thadius856** [comment]
> We will have to also pay a lot of attention when merging hotfixes. Hotfix updates are going to be based off of the release. There's a chance that the develop DB could look very different and the hotfix update could fail or completely FUBAR it.

Hotfixes are branched directly off the master. You sure this is a problem?

**Comment by JayUnt** [comment]
What if we have an xml file to associate sql files with releases. They can
be run in order it's in the file.  So when I'm done coding, I rebase, then
I throw my filename into the xml at the bottom. And then merge into master.

Thoughts?

**Comment by kerber** [comment]
> Hotfixes are branched directly off the master. You sure this is a problem?

Well, develop has the potential to be pretty different than master, right? As long as those two are different, it's going to require some extra work.

**Comment by thadius856** [comment]
> Well, develop has the potential to be pretty different than master, right? As long as those two are different, it's going to require some extra work.

As a matter of practice, develop should be rebased off the new master with each release. So while the develop branch may be different, it already includes all of the commits that master does, plus some more.

**Comment by JayUnt** [comment]
im not as worried with development, because the branches will be rebased off of develop, and then merged in. and then the same thing with develop and master when its time to release.

I'm worried that everything will work fine in develop/master, but then when it is released to the users, a script will get skipped and everything will break.  So we need a way to have the app know what needs to get run and in the correct order.

**Comment by Skemp45** [comment]
Closing if there is an issue in here you can open anew issue with the direct problem
